12/README.md

```diff
@@ -0,0 +1,165 @@
+
+EVENT_CHANNEL = "(upstream admin device FQDN)/(attr).EVENT_TYPE[_idl5]" 
+```
+
+## Runtime requirements
+
+Client and server are up and running. Server is reachable from client i.e. may communicate using Request-Reply protocol [RFC-10]. 
+
+## Publisher-Subscriber protocol
+
+![Subscription_1.png](images/Subscription_1.png)
+
+The main goal of the this protocol implementation is to allow client receive events from server:
+
+* Server MUST publish events of EVENT_TYPE
+
+* Server library MUST provide an API to publish events of EVENT_TYPE
+
+* Client SHOULD process events
+
+* Client library SHOULD provide user an API to register user defined event callbacks  
+
+### Negotiation
+
+![Subscription_2.png](images/Subscription_2.png)
+
+Negotiation phase implies that client executes admin's EventSubscriptionChange command:
+
+* Admin MUST implement EventSubscriptionChange command: 
+
+```ABNF
+EventSubscriptionChange = Tango Command
+EventSubscriptionChange_ARGIN-type  = DevVarStringArr 
+EventSubscriptionChange_ARGIN-value = "[ (upstream device), (attribute), "subscribe", EVENT_TYPE]"
+EventSubscriptionChange_ARGOUT-type = DevLong
+EventSubscriptionChange_ARGOUT-value = "(server library release ver)"
+```
+
+**NOTE**: Tango Server Library v9.3.3 force client to explicitly call ZmqEventSubscriptionChange to use ZMQ based implementation [RFC-13]: 
+ 
+```ABNF
+ZmqEventSubscriptionChange = Tango Command
+ZmqEventSubscriptionChangeArgument-type  = DevVarStringArr 
+ZmqEventSubscriptionChangeArgument-value = "[ upstream device, attribute, action ("subscribe|unsubscribe"), EVENT_TYPE]"
+ZmqEventSubscriptionChangeOutput-type = DevVarLongStringArr
+ZmqEventSubscriptionChangeOutput-value = "[(server library release ver), (upstream device idl ver), (zmq_sub_event_hwm), (rate), (ivl), (zmq release ver); (heartbeat_endpoint), (event_endpoint), n*(alternate_heartbeat_endpoint), n*(alternate_event_endpoint),(heartbeat_channel), (event_channel)]"
+```
+
+* Client MUST execute admin's EventSubscriptionChange command to establish Publisher-Subscriber relationship
+
+* Client SHOULD establish tcp? connection to heartbeat/event endpoints
+
+* Client MAY iterate over alternative heartbeat/event endpoints to establish connection if above has failed
+
+* Client MAY use provided heartbeat/event channels to identify event source 
+
+### Subscription 
+
+* Server MUST send events via corresponding EVENT_CHANNEL
+
+* Server MUST send HEARTBEAT event every 9 seconds
+
+* Client MUST execute EventConfirmSubscription command every 600 seconds to indicate its interest in subscription
+
+```abnf
+EventConfirmSubscription
+```  
+
+### Cancel event subscription
+
+* Client MAY cancel its subscription by executing EventSubscriptionChange on upstream admin 
+
+```abnf
+
+EventSubscriptionChange_ARGIN-value = "[ (upstream device), (attribute), "unsubscribe", EVENT_TYPE]"
+
+```
+
+* Server MUST stop to send events to client whose subscription is canceled
```

Comment:
Sorry if I bring some confusion. 
In the current implementation . When using ZMQ (at least) to send events, server MAY continue to send events to the client whose subscription is canceled if there is another client, interested in the same event type (same event type with same level of compatibility) which sent an event confirm subscription command in the last 599 seconds and if the client whose subscription is canceled didn't close the ZMQ event channel connection yet.

Reply:
@bourtemb This is interesting. Although I am not to follow the purpose. It is due to how Zmq works or the definition of what is a Client (computer / program / thread / socket ...)

Reply:
This is only due to the way the unsubscription is currently implemented. When a Tango client unsubscribes for a given event, there might be a short delay before the ZMQ unsubscription is actually sent on the network, so the server MIGHT continue to send some events until it really received the ZMQ unsubscription command.
The events received by the client between the unsubscribe_event() Tango call and the ZMQ event unsubscription reception on the server-side will simply be ignored by the Tango client.

Reply:
OK you describe also "another client"

Reply:
Right you are! What I wrote before is indeed confusing... But I warned at the beginning of the first comment: ;-)
> Sorry if I bring some confusion.

No need for "Another client"!